module/zfs/spa.c

```diff
@@ -22,7 +22,7 @@
 /*
  * Copyright (c) 2005, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 2011, 2017 by Delphix. All rights reserved.
- * Copyright (c) 2015, Nexenta Systems, Inc.  All rights reserved.
+ * Copyright (c) 2018, Nexenta Systems, Inc.  All rights reserved.
```

Comment:
Please merge the copyright with the next one on the next line.

Reply:
ok.

Regarding to DTL:

There is a pre-split check (see spa.c line 6466).

But the changed code is the final stage of split.
The record is a "stat" record (see vdev.c line 3870). So as I understand in this case we don't care about the record.

Reply:
> There is a pre-split check (see spa.c line 6466).

So the check is:
```
if (vdev_dtl_required(vml[c])) {
	error = SET_ERROR(EBUSY);
	break;
}
```
It appears that this check only makes sure that removing this vdev will not break the current pool, however it makes no checks for the new pool that we are splitting-out. Either I'm missing something and we make a similar check for the new pool elsewhere, or there is an issue with the current code.

Here's an example where the current code would trip:
```
mirror-1
    vdev-A (DTLs empty)
    vdev-B (has a missing/outage entry in DTL)
```

Say we split-out vdev-B into the new pool, the check above would pass, however the new pool will now have a top level vdev (vdev-B) that has DTLs (i.e. missing data) for some txgs.

Reply:
I just did some tests and it does appear that the scenario above results in a broken newpool.
Here are steps to reproduce this:
```
$ sudo zpool create p1 mirror xvdb xvdc
$ sudo zpool offline p1 xvdc
$ sudo dd if=/dev/urandom of=/p1/f1 bs=128k count=40000
$ sudo zpool online p1 xvdc; sleep 1; sudo zpool status p1
  pool: p1
 state: ONLINE
status: One or more devices is currently being resilvered.  The pool will
	continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Tue Sep  4 17:39:22 2018
	4.89G scanned at 1001M/s, 485M issued at 97.1M/s, 4.89G total
	477M resilvered, 9.70% done, 0 days 00:00:46 to go
config:

	NAME        STATE     READ WRITE CKSUM
	p1          ONLINE       0     0     0
	  mirror-0  ONLINE       0     0     0
	    xvdb    ONLINE       0     0     0
	    xvdc    ONLINE       0     0     0

errors: No known data errors
$ sudo zpool split p1 p2
$ sudo zpool status p1
  pool: p1
 state: ONLINE
status: One or more devices is currently being resilvered.  The pool will
	continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Tue Sep  4 17:39:22 2018
	4.89G scanned at 132M/s, 3.97G issued at 107M/s, 4.89G total
	1.86G resilvered, 81.31% done, 0 days 00:00:08 to go
config:

	NAME        STATE     READ WRITE CKSUM
	p1          ONLINE       0     0     0
	  xvdb      ONLINE       0     0     0

errors: No known data errors
$ sudo zpool import p2
$ sudo zpool scrub p2
$ sudo zpool status p2
  pool: p2
 state: DEGRADED
status: One or more devices has experienced an error resulting in data
	corruption.  Applications may be affected.
action: Restore the file in question if possible.  Otherwise restore the
	entire pool from backup.
   see: http://zfsonlinux.org/msg/ZFS-8000-8A
  scan: scrub repaired 0B in 0 days 00:01:06 with 24856 errors on Tue Sep  4 17:41:38 2018
config:

	NAME        STATE     READ WRITE CKSUM
	p2          DEGRADED     0     0 24.3K
	  xvdc      DEGRADED     0     0 48.5K  too many errors

errors: 24856 data errors, use '-v' for a list
```
We'll need to open a neb bug for this.

Reply:
> We'll need to open a new bug for this.
Do you mean don't allow to split if scrub/resilver in progress ?

Reply:
Scrub should not affect this, however if resilver is in-progress that'd likely mean that one of the vdevs in the mirror doesn't have data for some txgs and so we should fail the split.

Reply:
In the test I did above, if we offline the first vdev instead (`xvdb`), we will fail the split as expected, with:
```
$ sudo zpool split p1 p2
Unable to split p1: pool is busy
```
This most likely comes from the `vdev_dtl_required()` check.